doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -144,61 +144,129 @@ presentation of Cic is given in Chapter :ref:`calculusofinductiveconstructions`.
-                    : `term` :>
-                    : `term` -> `term`
-                    : `term` `arg` … `arg`
-                    : @ `qualid` [`term` … `term`]
-                    : `term` % `ident`
-                    : match `match_item` , … , `match_item` [`return_type`] with
-                    :   [[|] `equation` | … | `equation`] end
-                    : `qualid`
-                    : `sort`
-                    : `num`
-                    : _
-                    : ( `term` )
-   arg              : `term`
-                    : ( `ident` := `term` )
-   binders          : `binder` … `binder`
-   binder           : `name`
-                    : ( `name` … `name` : `term` )
-                    : ( `name` [: `term`] := `term` )
-                    : ' `pattern`
-   name             : `ident` | _
-   qualid           : `ident` | `qualid` `field`
-   sort             : SProp | Prop | Set | Type
-   fix_bodies       : `fix_body`
-                    : `fix_body` with `fix_body` with … with `fix_body` for `ident`
-   cofix_bodies     : `cofix_body`
-                    : `cofix_body` with `cofix_body` with … with `cofix_body` for `ident`
-   fix_body         : `ident` `binders` [`annotation`] [: `term`] := `term`
-   cofix_body       : `ident` [`binders`] [: `term`] := `term`
-   annotation       : { struct `ident` }
-   match_item       : `term` [as `name`] [in `qualid` [`pattern` … `pattern`]]
-   dep_ret_type     : [as `name`] `return_type`
-   return_type      : return `term`
-   equation         : `mult_pattern` | … | `mult_pattern` => `term`
-   mult_pattern     : `pattern` , … , `pattern`
-   pattern          : `qualid` `pattern` … `pattern`
-                    : @ `qualid` `pattern` … `pattern`
-                    : `pattern` as `ident`
-                    : `pattern` % `ident`
-                    : `qualid`
-                    : _
-                    : `num`
-                    : ( `pattern` | … | `pattern` )
+   term : `binder_constr`
+        : `term100`
+   term100 : `term_cast`
+           : `term10` :>
+           : `term10`
+   term10 : `term9` `appl_arg_list`
+          : @ `qualid` `instance` `term9_list_opt`
+          : @ ? `ident` `ident_list`
+          : `term9`
+   ident_list : `ident_list` `ident`
+              : `ident`
+   appl_arg_list : `appl_arg_list` `appl_arg`
+                 : `appl_arg`
+   appl_arg : ( `ident` := `term` )
+            : `term9`
+   term9 : .. `term0` ..
+         : `term1`
+   term1 : `term0` .( `qualid` `appl_arg_list_opt` )
+         : `term0` .( @ `qualid` `term9_list_opt` )
+         : `term0` % `ident`
+         : `term0`
+   appl_arg_list_opt : `appl_arg_list_opt` `appl_arg`
+                     : `empty`
+   term9_list_opt : `term9_list_opt` `term9`
+                  : `empty`
+   empty :
+   name_colon : `ident` :
+              : _ :
+   term0 : `atomic_constr`
+         : `match_constr`
+         : ( `term` )
+         : {| `record_fields` | }
```

Comment:
The spacing is weird. According to this rule, a space between the final `|` and `}` would be OK, but not at the beginning. I propose to pretend this final space is not allowed either.

Reply:
Weird but harmless, right?.  `| }` appears in 3 productions.  I think there are some other similar cases.  Of course I'll change it if you still think it has to.

Reply:
If it is easy to change, it would simply make the reading easier. Thus, I would not say it is completely harmless to keep as is. If it is too complicated, do not bother as this is only a minor point.

Reply:
Done.

Here are all the other orderedGrammar  productions with adjacent non-alphabetic terminals.  You want to change any of these too?

```
| "(" "_" ":" term "|" term ")"
| "@" "?" ident ident_list
| term0 ".(" "@" qualid term9_list_opt ")"
| "?" "[" ident "]"
| "?" "[" "?" ident "]"
| "{|" record_fields "|" "}"
| "ltac" ":" "(" ltac_expr ")"
| "{" name "}" ":" exclam_opt term
| "_" ":"
| ":>" ">"
| ":" ">" ">"
| ":" ">"
| "Type" "@{" "_" "}"
| "{|" record_patterns "|" "}"
| "-" "[" smart_global_list "]"
| ":=" "{" record_fields "}"
| "(" starredidentref_list_opt ")" "*"
| "(" ssexpr ")" "*"
| "in" "|-" "*"
| "*" "|-" concl_occ
| "_" ":=" ltac_expr
| ltac_expr3 ";" "[" multi_goal_tactics "]"
| ltac_expr3 ";" "[" ">" multi_goal_tactics "]"
| "[" ">" multi_goal_tactics "]"
| "!" ":"
| "[" match_hyp_list_comma_opt "|-" match_pattern "]" "=>" ltac_expr
| "_" "=>" ltac_expr
| "[" match_pattern "]" ":"
```

Reply:
Done.

I'd do these, too:

```
of_type_with_opt_coercion: [
| ":>>"
| ":>" ">"
| ":>"
| ":" ">" ">"
| ":" ">"
| ":"
]
```

BTW, I think you're looking for "asymmetry".

Reply:
Agreed.